File: plugin/go/contract/contract.go
package contract

import (
	"bytes"
	"encoding/binary"
	"log"
	"math"
	"math/rand"

"google.golang.org/protobuf/types/known/anypb"
)

var (
PREFIX_MARKET_STATE    = []byte{0x10}
PREFIX_POSITION_STATE  = []byte{0x11}
PREFIX_OUTCOME_STATE   = []byte{0x12}
PREFIX_RESOLVER_STATE  = []byte{0x13}
PREFIX_TREASURY        = []byte{0x14}
PREFIX_RESOLVER_RECORD = []byte{0x16}
PREFIX_PROPOSAL_RECORD = []byte{0x17}
PREFIX_DISPUTE_RECORD  = []byte{0x18}
PREFIX_VOTE_COMMIT     = []byte{0x19}
PREFIX_VOTE_REVEAL     = []byte{0x1A}
PREFIX_SLASH_RECORD    = []byte{0x1B}
)

var (
PREFIX_ACCOUNT  = []byte{0x01}
PREFIX_FEE_POOL = []byte{0x02}
)

// Issue-12: assert MIN_B0 > FINALIZATION_BOUNTY at startup.
// If this ever fails, create_market would seed a TreasuryReserve that cannot
// cover the finalization bounty, silently breaking permissionless finalization.
func init() {
if MIN_B0 <= FINALIZATION_BOUNTY {
panic("invariant violated: MIN_B0 must be greater than FINALIZATION_BOUNTY")
}
}

var ContractConfig = &PluginConfig{
Name:    "praxis_prediction_market",
Id:      1,
Version: 1,
SupportedTransactions: []string{
"create_market",
"submit_prediction",
"claim_winnings",
"register_resolver",
"propose_outcome",
"file_dispute",
"commit_vote",
"reveal_vote",
"tally_votes",
"finalize_market",
"claim_slash",
"reclaim_stake",
"forfeit_position",
	"claim_builder_reward",
	"claim_creator_fee",
	"claim_resolver_reward",
		"claim_community_reward",
		"claim_investor_reward",
		"claim_protocol_reward",
		"unstake_resolver",
		"cancel_market",
		"claim_unbonded_stake",
},
TransactionTypeUrls: []string{
	"type.googleapis.com/types.MessageCreateMarket",
	"type.googleapis.com/types.MessageSubmitPrediction",
	"type.googleapis.com/types.MessageClaimWinnings",
	"type.googleapis.com/types.MessageRegisterResolver",
	"type.googleapis.com/types.MessageProposeOutcome",
	"type.googleapis.com/types.MessageFileDispute",
	"type.googleapis.com/types.MessageCommitVote",
	"type.googleapis.com/types.MessageRevealVote",
	"type.googleapis.com/types.MessageTallyVotes",
	"type.googleapis.com/types.MessageFinalizeMarket",
	"type.googleapis.com/types.MessageClaimSlash",
	"type.googleapis.com/types.MessageReclaimStake",
	"type.googleapis.com/types.MessageForfeitPosition",
	"type.googleapis.com/types.MessageClaimBuilderReward",
	"type.googleapis.com/types.MessageClaimCreatorFee",
	"type.googleapis.com/types.MessageClaimResolverReward",
		"type.googleapis.com/types.MessageClaimCommunityReward",
		"type.googleapis.com/types.MessageClaimInvestorReward",
		"type.googleapis.com/types.MessageClaimProtocolReward",
		"type.googleapis.com/types.MessageUnstakeResolver",
		"type.googleapis.com/types.MessageCancelMarket",
		"type.googleapis.com/types.MessageClaimUnbondedStake",
},
}



type Contract struct {
Config    Config
FSMConfig *PluginFSMConfig
plugin    *Plugin
fsmId     uint64
}

func marketKey(prefix, marketId []byte) []byte {
return append(append([]byte{}, prefix...), marketId...)
}
func positionKey(marketId, addr []byte) []byte {
k := append(append([]byte{}, PREFIX_POSITION_STATE...), marketId...)
return append(k, addr...)
}
func addrKey(prefix, addr []byte) []byte {
return append(append([]byte{}, prefix...), addr...)
}

func (c *Contract) Genesis(req *PluginGenesisRequest) *PluginGenesisResponse {
return &PluginGenesisResponse{}
}

func (c *Contract) BeginBlock(req *PluginBeginRequest) *PluginBeginResponse {
SetGlobalHeight(req.Height)

entropyQId := nextQueryId()
entropyResp, readErr := c.plugin.StateRead(c, &PluginStateReadRequest{
Keys: []*PluginKeyRead{
{QueryId: entropyQId, Key: PANEL_ENTROPY_KEY},
},
})

// Issue-17 fix: replace XOR accumulator with SHA256 hash chain.
// XOR with a deterministic height function is fully predictable by chain
// observers — they can compute the accumulator value at any future block and
// time a file_dispute call to influence panel selection.
// SHA256(prev || height_bytes) is a one-way function: knowing the output
// does not allow computing a height that produces a desired panel seed.
var prev [8]byte
if readErr == nil && entropyResp != nil {
for _, r := range entropyResp.Results {
if r.QueryId == entropyQId && len(r.Entries) > 0 && len(r.Entries[0].Value) >= 8 {
copy(prev[:], r.Entries[0].Value[:8])
}
}
}

// DeliverTx() is code that is executed to apply a transaction
func (c *Contract) DeliverTx(request *PluginDeliverRequest) *PluginDeliverResponse {
	// get the message
	msg, err := FromAny(request.Tx.Msg)
	if err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	// handle the message
	switch x := msg.(type) {
	case *MessageSend:
		return c.DeliverMessageSend(x, request.Tx.Fee, request.Tx.Memo)
	default:
		return &PluginDeliverResponse{Error: ErrInvalidMessageCast()}
	}
}
input := append(prev[:], heightBytes...)
hash := sha256.Sum256(input)
buf := hash[:8]

wr, writeErr := c.plugin.StateWrite(c, &PluginStateWriteRequest{
Sets: []*PluginSetOp{
{Key: PANEL_ENTROPY_KEY, Value: buf},
},
})
if writeErr != nil || (wr != nil && wr.Error != nil) {
_ = wr
}

return &PluginBeginResponse{}
}

// DeliverMessageSend() handles a 'send' message
func (c *Contract) DeliverMessageSend(msg *MessageSend, fee uint64, memo string) *PluginDeliverResponse {
	log.Printf("DeliverMessageSend called: from=%x to=%x amount=%d fee=%d", msg.FromAddress, msg.ToAddress, msg.Amount, fee)
	var (
		fromKey, toKey, feePoolKey         []byte
		fromBytes, toBytes, feePoolBytes   []byte
		fromQueryId, toQueryId, feeQueryId = rand.Uint64(), rand.Uint64(), rand.Uint64()
		from, to, feePool                  = new(Account), new(Account), new(Pool)
	)
	// calculate the from key and to key
	fromKey, toKey, feePoolKey = KeyForAccount(msg.FromAddress), KeyForAccount(msg.ToAddress), KeyForFeePool(c.Config.ChainId)
	log.Printf("Keys: fromKey=%x toKey=%x feePoolKey=%x", fromKey, toKey, feePoolKey)
	// get the from and to account
	response, err := c.plugin.StateRead(c, &PluginStateReadRequest{
		Keys: []*PluginKeyRead{
			{QueryId: feeQueryId, Key: feePoolKey},
			{QueryId: fromQueryId, Key: fromKey},
			{QueryId: toQueryId, Key: toKey},
		}})
	// check for internal error
	if err != nil {
		log.Printf("StateRead error: %v", err)
		return &PluginDeliverResponse{Error: err}
	}
	// ensure no error fsm error
	if response.Error != nil {
		log.Printf("StateRead FSM error: %v", response.Error)
		return &PluginDeliverResponse{Error: response.Error}
	}
	log.Printf("StateRead returned %d results", len(response.Results))
	// get the from bytes and to bytes
	for _, resp := range response.Results {
		log.Printf("Result QueryId=%d Entries=%d", resp.QueryId, len(resp.Entries))
		if len(resp.Entries) == 0 {
			log.Printf("WARNING: No entries for QueryId=%d", resp.QueryId)
			continue
		}
		switch resp.QueryId {
		case fromQueryId:
			fromBytes = resp.Entries[0].Value
			log.Printf("fromBytes len=%d", len(fromBytes))
		case toQueryId:
			toBytes = resp.Entries[0].Value
			log.Printf("toBytes len=%d", len(toBytes))
		case feeQueryId:
			feePoolBytes = resp.Entries[0].Value
			log.Printf("feePoolBytes len=%d", len(feePoolBytes))
		}
	}
	// convert the bytes to account structures
	if err = Unmarshal(fromBytes, from); err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	if err = Unmarshal(toBytes, to); err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	if err = Unmarshal(feePoolBytes, feePool); err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	if msg.Amount > math.MaxUint64-fee {
		return &PluginDeliverResponse{Error: ErrInvalidAmount()}
	}
	amountToDeduct := msg.Amount + fee
	log.Printf("from.Amount=%d to.Amount=%d feePool.Amount=%d", from.Amount, to.Amount, feePool.Amount)
	// if the account amount is less than the amount to subtract; return insufficient funds
	if from.Amount < amountToDeduct {
		log.Printf("ERROR: Insufficient funds: from.Amount=%d amountToDeduct=%d", from.Amount, amountToDeduct)
		return &PluginDeliverResponse{Error: ErrInsufficientFunds()}
	}
	// for self-transfer, use same account data
	isSelfTransfer := bytes.Equal(fromKey, toKey)
	if isSelfTransfer {
		to = from
	}
	if feePool.Amount > math.MaxUint64-fee || (!isSelfTransfer && to.Amount > math.MaxUint64-msg.Amount) {
		return &PluginDeliverResponse{Error: ErrInvalidAmount()}
	}
	// subtract from sender
	from.Amount -= amountToDeduct
	// add the fee to the 'fee pool'
	feePool.Amount += fee
	// add to recipient
	to.Amount += msg.Amount
	log.Printf("AFTER: from.Amount=%d to.Amount=%d feePool.Amount=%d", from.Amount, to.Amount, feePool.Amount)
	// convert the accounts to bytes
	fromBytes, err = Marshal(from)
	if err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	toBytes, err = Marshal(to)
	if err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	feePoolBytes, err = Marshal(feePool)
	if err != nil {
		return &PluginDeliverResponse{Error: err}
	}
	// Retain drained accounts only when they carry nonce state or core will advance the nonce after RLP.V2 delivery.
	request := &PluginStateWriteRequest{Sets: []*PluginSetOp{
		{Key: feePoolKey, Value: feePoolBytes},
		{Key: toKey, Value: toBytes},
	}}
	if from.Amount == 0 && from.Nonce == 0 && memo != "RLP.V2" {
		request.Deletes = []*PluginDeleteOp{{Key: fromKey}}
	} else {
		request.Sets = append(request.Sets, &PluginSetOp{Key: fromKey, Value: fromBytes})
	}
	resp, err := c.plugin.StateWrite(c, request)
	if err != nil {
		log.Printf("StateWrite internal error: %v", err)
		return &PluginDeliverResponse{Error: err}
	}
	if resp.Error != nil {
		log.Printf("StateWrite FSM error: %v", resp.Error)
		return &PluginDeliverResponse{Error: resp.Error}
	}
	log.Printf("StateWrite SUCCESS!")
	return &PluginDeliverResponse{}
}

func (c *Contract) CheckTx(req *PluginCheckRequest) *PluginCheckResponse {
msg, err := FromAny(req.Tx.Msg)
if err != nil {
return &PluginCheckResponse{Error: err}
}
switch m := msg.(type) {
case *MessageSend:
return c.CheckMessageSend(m)
case *MessageCreateMarket:
return c.CheckMessageCreateMarket(m)
case *MessageSubmitPrediction:
return c.CheckMessageSubmitPrediction(m)
case *MessageClaimWinnings:
return c.CheckMessageClaimWinnings(m)
case *MessageRegisterResolver:
return c.CheckMessageRegisterResolver(m)
case *MessageProposeOutcome:
return c.CheckMessageProposeOutcome(m)
case *MessageFileDispute:
return c.CheckMessageFileDispute(m)
case *MessageCommitVote:
return c.CheckMessageCommitVote(m)
case *MessageRevealVote:
return c.CheckMessageRevealVote(m)
case *MessageTallyVotes:
return c.CheckMessageTallyVotes(m)
case *MessageFinalizeMarket:
return c.CheckMessageFinalizeMarket(m)
case *MessageClaimSlash:
return c.CheckMessageClaimSlash(m)
case *MessageReclaimStake:
return c.CheckMessageReclaimStake(m)
case *MessageForfeitPosition:
return c.CheckMessageForfeitPosition(m)
case *MessageClaimBuilderReward:
return c.CheckMessageClaimBuilderReward(m)
case *MessageClaimCreatorFee:
return c.CheckMessageClaimCreatorFee(m)
case *MessageClaimResolverReward:
return c.CheckMessageClaimResolverReward(m)
case *MessageClaimCommunityReward:
return c.CheckMessageClaimCommunityReward(m)
case *MessageClaimInvestorReward:
return c.CheckMessageClaimInvestorReward(m)
case *MessageClaimProtocolReward:
return c.CheckMessageClaimProtocolReward(m)
case *MessageUnstakeResolver:
return c.CheckMessageUnstakeResolver(m)
case *MessageCancelMarket:
return c.CheckMessageCancelMarket(m)
case *MessageClaimUnbondedStake:
return c.CheckMessageClaimUnbondedStake(m)
default:
return &PluginCheckResponse{Error: ErrInvalidMessageCast()}
}
}

func (c *Contract) DeliverTx(req *PluginDeliverRequest) *PluginDeliverResponse {
msg, err := FromAny(req.Tx.Msg)
if err != nil {
return &PluginDeliverResponse{Error: err}
}
fee := req.Tx.Fee
switch m := msg.(type) {
case *MessageSend:
return c.DeliverMessageSend(m, fee)
case *MessageCreateMarket:
return c.DeliverMessageCreateMarket(m, fee)
case *MessageSubmitPrediction:
return c.DeliverMessageSubmitPrediction(m, fee)
case *MessageClaimWinnings:
return c.DeliverMessageClaimWinnings(m, fee)
case *MessageRegisterResolver:
return c.DeliverMessageRegisterResolver(m, fee)
case *MessageProposeOutcome:
return c.DeliverMessageProposeOutcome(m, fee)
case *MessageFileDispute:
return c.DeliverMessageFileDispute(m, fee)
case *MessageCommitVote:
return c.DeliverMessageCommitVote(m, fee)
case *MessageRevealVote:
return c.DeliverMessageRevealVote(m, fee)
case *MessageTallyVotes:
return c.DeliverMessageTallyVotes(m, fee)
case *MessageFinalizeMarket:
return c.DeliverMessageFinalizeMarket(m, fee)
case *MessageClaimSlash:
return c.DeliverMessageClaimSlash(m, fee)
case *MessageReclaimStake:
return c.DeliverMessageReclaimStake(m, fee)
case *MessageForfeitPosition:
return c.DeliverMessageForfeitPosition(m, fee)
case *MessageClaimBuilderReward:
return c.DeliverMessageClaimBuilderReward(m, fee)
case *MessageClaimCreatorFee:
return c.DeliverMessageClaimCreatorFee(m, fee)
case *MessageClaimResolverReward:
return c.DeliverMessageClaimResolverReward(m, fee)
case *MessageClaimCommunityReward:
return c.DeliverMessageClaimCommunityReward(m, fee)
case *MessageClaimInvestorReward:
return c.DeliverMessageClaimInvestorReward(m, fee)
case *MessageClaimProtocolReward:
return c.DeliverMessageClaimProtocolReward(m, fee)
case *MessageUnstakeResolver:
return c.DeliverMessageUnstakeResolver(m, fee)
case *MessageCancelMarket:
return c.DeliverMessageCancelMarket(m, fee)
case *MessageClaimUnbondedStake:
return c.DeliverMessageClaimUnbondedStake(m, fee)
default:
return &PluginDeliverResponse{Error: ErrInvalidMessageCast()}
}
}

var _ = (*anypb.Any)(nil)

File: plugin/python/contract/contract.py
"""
Contract implementation for Canopy blockchain plugin.

This file contains the base contract implementation that handles the 'send' transaction.
Matches Go's contract/contract.go structure.
"""

import json
import random
import struct
from typing import Optional, Dict, Any, Union, Protocol, TYPE_CHECKING

UINT64_MAX = (1 << 64) - 1

if TYPE_CHECKING:
    from .plugin import Plugin, Config

# Import proto types
from .proto import (
    PluginCheckRequest,
    PluginCheckResponse,
    PluginDeliverRequest,
    PluginDeliverResponse,
    PluginGenesisRequest,
    PluginGenesisResponse,
    PluginBeginRequest,
    PluginBeginResponse,
    PluginEndRequest,
    PluginEndResponse,
    MessageSend,
    MessagePredict,
    PluginKeyRead,
    PluginStateReadRequest,
    PluginStateWriteRequest,
    PluginSetOp,
    PluginDeleteOp,
    PluginFSMConfig,
    FeeParams,
    Account,
    Pool,
)
from .proto import account_pb2, event_pb2, plugin_pb2, tx_pb2
from google.protobuf import any_pb2
from .ai_model import get_model

from .error import (
    PluginError,
    err_invalid_address,
    err_invalid_amount,
    err_insufficient_funds,
    err_tx_fee_below_state_limit,
    err_invalid_message_cast,
    err_unmarshal,
)


# Plugin configuration (matching Go's ContractConfig)
CONTRACT_CONFIG = {
    "name": "python_plugin_contract",
    "id": 1,
    "version": 1,
    "supported_transactions": ["send", "predict"],
    "transaction_type_urls": [
        "type.googleapis.com/types.MessageSend",
        "type.googleapis.com/types.MessagePredict",
    ],
    "event_type_urls": [],
    # Include google/protobuf/any.proto first as it's a dependency of event.proto and tx.proto
    "file_descriptor_protos": [
        any_pb2.DESCRIPTOR.serialized_pb,
        account_pb2.DESCRIPTOR.serialized_pb,
        event_pb2.DESCRIPTOR.serialized_pb,
        plugin_pb2.DESCRIPTOR.serialized_pb,
        tx_pb2.DESCRIPTOR.serialized_pb,
    ],
}


# State key prefixes (matching Go)
ACCOUNT_PREFIX = b"\x01"
POOL_PREFIX = b"\x02"
# Predict results (AI/ML inference output from G2 ZeroPerceptron)
PREDICT_PREFIX = b"\x03"
PARAMS_PREFIX = b"\x07"


# Key generation functions (from keys.py)

def join_len_prefix(*items: Optional[bytes]) -> bytes:
    """Join byte arrays with length prefixes."""
    result = bytearray()
    for item in items:
        if not item:
            continue
        if len(item) > 255:
            raise ValueError(f"Item too long: {len(item)} bytes (max 255)")
        result.append(len(item))
        result.extend(item)
    return bytes(result)


def format_uint64(value: Union[int, str]) -> bytes:
    """Format uint64 as big-endian bytes."""
    if isinstance(value, str):
        value = int(value)
    if not isinstance(value, int) or value < 0 or value >= (1 << 64):
        raise ValueError(f"Invalid uint64 value: {value}")
    return struct.pack('>Q', value)


def key_for_account(address: bytes) -> bytes:
    """Generate state database key for an account."""
    return join_len_prefix(ACCOUNT_PREFIX, address)


def key_for_fee_params() -> bytes:
    """Generate state database key for fee parameters."""
    return join_len_prefix(PARAMS_PREFIX, b"/f/")


def key_for_fee_pool(chain_id: int) -> bytes:
    """Generate state database key for fee pool."""
    return join_len_prefix(POOL_PREFIX, format_uint64(chain_id))


def key_for_predict_log(address: bytes, seq: int) -> bytes:
    """Generate state database key for an AI prediction result.

    Namespace: 0x03 + address + sequence (big-endian)."""
    return join_len_prefix(PREDICT_PREFIX, address, format_uint64(seq))


# Proto marshal/unmarshal utilities

def marshal(message: Any) -> bytes:
    """Marshal object to protobuf bytes."""
    try:
        if hasattr(message, 'SerializeToString'):
            return message.SerializeToString()
        raise ValueError("Message does not support serialization")
    except Exception as err:
        raise err_unmarshal(err)


def unmarshal(message_type: Any, data: Optional[bytes]) -> Optional[Any]:
    """Unmarshal bytes to protobuf message."""
    if not data:
        return None
    try:
        if hasattr(message_type, 'FromString'):
            return message_type.FromString(data)
        raise ValueError("Message type does not support deserialization")
    except Exception as err:
        raise err_unmarshal(err)


class Contract:
    """
    Contract defines the smart contract that implements the extended logic of the nested chain.
    Matches Go's Contract struct.
    """

    def __init__(
        self,
        config: Optional["Config"] = None,
        fsm_config: Optional[PluginFSMConfig] = None,
        plugin: Optional["Plugin"] = None,
        fsm_id: Optional[int] = None,
    ):
        self.config = config
        self.fsm_config = fsm_config
        self.plugin = plugin
        self.fsm_id = fsm_id

    def genesis(self, request: PluginGenesisRequest) -> PluginGenesisResponse:
        """Genesis implements logic to import a json file to create the state at height 0."""
        return PluginGenesisResponse()

    def begin_block(self, request: PluginBeginRequest) -> PluginBeginResponse:
        """BeginBlock is code that is executed at the start of applying the block."""
        return PluginBeginResponse()

    async def check_tx(self, request: PluginCheckRequest) -> PluginCheckResponse:
        """CheckTx is code that is executed to statelessly validate a transaction."""
        try:
            if not self.plugin or not self.config:
                raise PluginError(1, "plugin", "plugin or config not initialized")

            # Validate fee - read fee params from state
            resp = await self.plugin.state_read(
                self,
                PluginStateReadRequest(
                    keys=[PluginKeyRead(query_id=random.randint(0, 2**53), key=key_for_fee_params())]
                ),
            )

            if resp.HasField("error"):
                response = PluginCheckResponse()
                response.error.CopyFrom(resp.error)
                return response

            # Convert bytes into fee parameters
            if not resp.results or not resp.results[0].entries:
                raise PluginError(1, "plugin", "Fee parameters not found")

            fee_params_bytes = resp.results[0].entries[0].value
            min_fees = unmarshal(FeeParams, fee_params_bytes)
            if not min_fees:
                raise PluginError(1, "plugin", "Failed to decode fee parameters")

            # Get the message and handle by type
            type_url = request.tx.msg.type_url

            # Check for minimum fee (predict uses its own fee threshold)
            if type_url.endswith("/types.MessagePredict"):
                if request.tx.fee < min_fees.predict_fee:
                    raise err_tx_fee_below_state_limit()
            elif request.tx.fee < min_fees.send_fee:
                raise err_tx_fee_below_state_limit()

            if type_url.endswith("/types.MessageSend"):
                msg = MessageSend()
                msg.ParseFromString(request.tx.msg.value)
                return self._check_message_send(msg)
            elif type_url.endswith("/types.MessagePredict"):
                msg = MessagePredict()
                msg.ParseFromString(request.tx.msg.value)
                return self._check_message_predict(msg)
            else:
                raise err_invalid_message_cast()

        except PluginError as e:
            response = PluginCheckResponse()
            response.error.code = e.code
            response.error.module = e.module
            response.error.msg = e.msg
            return response
        except Exception as err:
            response = PluginCheckResponse()
            response.error.code = 1
            response.error.module = "plugin"
            response.error.msg = str(err)
            return response

    async def deliver_tx(self, request: PluginDeliverRequest) -> PluginDeliverResponse:
        """DeliverTx is code that is executed to apply a transaction."""
        try:
            # Get the message and handle by type
            type_url = request.tx.msg.type_url
            if type_url.endswith("/types.MessageSend"):
                msg = MessageSend()
                msg.ParseFromString(request.tx.msg.value)
                return await self._deliver_message_send(msg, request.tx.fee, request.tx.memo)
            elif type_url.endswith("/types.MessagePredict"):
                msg = MessagePredict()
                msg.ParseFromString(request.tx.msg.value)
                return await self._deliver_message_predict(msg, request.tx.fee, request.tx.memo)
            else:
                raise err_invalid_message_cast()

        except PluginError as e:
            response = PluginDeliverResponse()
            response.error.code = e.code
            response.error.module = e.module
            response.error.msg = e.msg
            return response
        except Exception as err:
            response = PluginDeliverResponse()
            response.error.code = 1
            response.error.module = "plugin"
            response.error.msg = str(err)
            return response

    def end_block(self, request: PluginEndRequest) -> PluginEndResponse:
        """EndBlock is code that is executed at the end of applying a block."""
        return PluginEndResponse()

    def _check_message_send(self, msg: MessageSend) -> PluginCheckResponse:
        """CheckMessageSend statelessly validates a 'send' message."""
        # Check sender address (must be exactly 20 bytes)
        if len(msg.from_address) != 20:
            raise err_invalid_address()

        # Check recipient address (must be exactly 20 bytes)
        if len(msg.to_address) != 20:
            raise err_invalid_address()

        # Check amount (must be greater than 0)
        if msg.amount == 0:
            raise err_invalid_amount()

        # Return authorized signers (sender must sign)
        response = PluginCheckResponse()
        response.recipient = msg.to_address
        response.authorized_signers.append(msg.from_address)
        return response

    async def _deliver_message_send(self, msg: MessageSend, fee: int, memo: str) -> PluginDeliverResponse:
        """DeliverMessageSend handles a 'send' message."""
        if not self.plugin or not self.config:
            raise PluginError(1, "plugin", "plugin or config not initialized")

        # Generate query IDs
        from_query_id = random.randint(0, 2**53)
        to_query_id = random.randint(0, 2**53)
        fee_query_id = random.randint(0, 2**53)

        # Calculate keys
        from_key = key_for_account(msg.from_address)
        to_key = key_for_account(msg.to_address)
        fee_pool_key = key_for_fee_pool(self.config.chain_id)

        # Get the from and to accounts
        response = await self.plugin.state_read(
            self,
            PluginStateReadRequest(
                keys=[
                    PluginKeyRead(query_id=fee_query_id, key=fee_pool_key),
                    PluginKeyRead(query_id=from_query_id, key=from_key),
                    PluginKeyRead(query_id=to_query_id, key=to_key),
                ]
            ),
        )

        # Check for internal error
        if response.HasField("error"):
            result = PluginDeliverResponse()
            result.error.CopyFrom(response.error)
            return result

        # Get the from bytes and to bytes
        from_bytes = None
        to_bytes = None
        fee_pool_bytes = None

        for resp in response.results:
            if resp.query_id == from_query_id:
                from_bytes = resp.entries[0].value if resp.entries else None
            elif resp.query_id == to_query_id:
                to_bytes = resp.entries[0].value if resp.entries else None
            elif resp.query_id == fee_query_id:
                fee_pool_bytes = resp.entries[0].value if resp.entries else None

        if msg.amount > UINT64_MAX - fee:
            raise err_invalid_amount()

        # Add fee to amount to deduct
        amount_to_deduct = msg.amount + fee

        # Convert bytes to account structures
        from_account = unmarshal(Account, from_bytes) if from_bytes else Account()
        to_account = unmarshal(Account, to_bytes) if to_bytes else Account()
        fee_pool = unmarshal(Pool, fee_pool_bytes) if fee_pool_bytes else Pool()

        # Check sufficient funds
        if from_account.amount < amount_to_deduct:
            raise err_insufficient_funds()

        # For self-transfer, use same account data
        if from_key == to_key:
            to_account = from_account

        if fee_pool.amount > UINT64_MAX - fee or (
            from_key != to_key and to_account.amount > UINT64_MAX - msg.amount
        ):
            raise err_invalid_amount()

        # Subtract from sender
        from_account.amount -= amount_to_deduct

        # Add the fee to the fee pool
        fee_pool.amount += fee

        # Add to recipient
        to_account.amount += msg.amount

        # Convert accounts to bytes
        from_bytes_new = marshal(from_account)
        to_bytes_new = marshal(to_account)
        fee_pool_bytes_new = marshal(fee_pool)

        # Retain drained accounts only when they carry nonce state or core will advance the nonce after RLP.V2 delivery.
        sets = [
            PluginSetOp(key=fee_pool_key, value=fee_pool_bytes_new),
            PluginSetOp(key=to_key, value=to_bytes_new),
        ]
        deletes = []
        if from_account.amount == 0 and from_account.nonce == 0 and memo != "RLP.V2":
            deletes.append(PluginDeleteOp(key=from_key))
        else:
            sets.append(PluginSetOp(key=from_key, value=from_bytes_new))
        write_resp = await self.plugin.state_write(
            self,
            PluginStateWriteRequest(
                sets=sets,
                deletes=deletes,
            ),
        )

        result = PluginDeliverResponse()
        if write_resp.HasField("error"):
            result.error.CopyFrom(write_resp.error)
        return result

    def _check_message_predict(self, msg: MessagePredict) -> PluginCheckResponse:
        """CheckMessagePredict statelessly validates a 'predict' message.

        The message carries multimodal features (text / numbers / timestamp)
        that the G2 ZeroPerceptron model consumes on-chain."""
        # Check sender address (must be exactly 20 bytes)
        if len(msg.from_address) != 20:
            raise err_invalid_address()

        # Features may be empty (pure text-event) but never negative:
        # model normalizes it. We only reject absurd sizes to keep
        # transaction size bounded.
        if len(msg.features) > 256:
            raise PluginError(1, "plugin", "features too large")

        # Return authorized signers (sender must sign)
        response = PluginCheckResponse()
        response.authorized_signers.append(msg.from_address)
        return response

    async def _deliver_message_predict(self, msg: MessagePredict, fee: int, memo: str) -> PluginDeliverResponse:
        """DeliverMessagePredict runs on-chain AI inference.

        Features from the message are passed through the G2 28D encoder
        (text/numbers/timestamp modalities with octonion cross7 fusion)
        and the ZeroPerceptron backbone produces a 16-class decision.

        The result (class + confidence) is written into state under
        PREDICT_PREFIX (b'\\x03') keyed by sender address + nonce.
        """
        if not self.plugin or not self.config:
            raise PluginError(1, "plugin", "plugin or config not initialized")

        # Read sender account (for fee deduction + nonce)
        from_key = key_for_account(msg.from_address)
        fee_pool_key = key_for_fee_pool(self.config.chain_id)
        from_query_id = random.randint(0, 2**53)
        fee_query_id = random.randint(0, 2**53)

        response = await self.plugin.state_read(
            self,
            PluginStateReadRequest(
                keys=[
                    PluginKeyRead(query_id=fee_query_id, key=fee_pool_key),
                    PluginKeyRead(query_id=from_query_id, key=from_key),
                ]
            ),
        )
        if response.HasField("error"):
            result = PluginDeliverResponse()
            result.error.CopyFrom(response.error)
            return result

        from_bytes = None
        fee_pool_bytes = None
        for resp in response.results:
            if resp.query_id == from_query_id:
                from_bytes = resp.entries[0].value if resp.entries else None
            elif resp.query_id == fee_query_id:
                fee_pool_bytes = resp.entries[0].value if resp.entries else None

        from_account = unmarshal(Account, from_bytes) if from_bytes else Account()
        fee_pool = unmarshal(Pool, fee_pool_bytes) if fee_pool_bytes else Pool()

        # Must be able to pay the predict fee
        if from_account.amount < fee:
            raise err_insufficient_funds()

        # Run the G2 ZeroPerceptron inference.
        # MessagePredict carries only numeric features; a 28-D vector is
        # consumed directly (raw mode). Shorter vectors (7-D numeric or
        # empty) are encoded via the G2 multimodal numeric encoder path.
        model = get_model()
        model.reset()
        features = list(msg.features)
        if len(features) == 28:
            result = model.predict_raw(features)
        else:
            result = model.predict_from_event(
                text="",
                numbers=features,
                timestamp=0.0,
            )

        # Deduct fee -> pool, keep nonce
        from_account.amount -= fee
        fee_pool.amount += fee

        # Persist prediction log under 0x03 namespace
        seq = from_account.nonce
        predict_key = key_for_predict_log(msg.from_address, seq)
        log_value = json.dumps({
            "y": result["y"],
            "class": result.get("class", ""),
            "top_probs": result.get("probs", [])[:3],
            "top3": result.get("top3", []),
            "feature_importance": result.get("feature_importance", []),
            "temperature": result.get("temperature", 1.5),
            "height": 0,
        }).encode("utf-8")

        sets = [
            PluginSetOp(key=from_key, value=marshal(from_account)),
            PluginSetOp(key=fee_pool_key, value=marshal(fee_pool)),
            PluginSetOp(key=predict_key, value=log_value),
        ]
        write_resp = await self.plugin.state_write(
            self,
            PluginStateWriteRequest(sets=sets, deletes=[]),
        )

        result = PluginDeliverResponse()
        result.events.extend([])
        if write_resp.HasField("error"):
            result.error.CopyFrom(write_resp.error)
        return result
